File: py/src/imbrace/resources/marketplace.py
from typing import Any, Dict, List, Optional, Union
from ..http import HttpTransport, AsyncHttpTransport
from ..types.marketplace import (
    Product, Order, CreateOrderInput, OrderStatus,
    EmailTemplate, CreateEmailTemplateInput,
    MarketplaceFileDetails, MarketplaceFileUploadResponse, MarketplaceCategory
)
from ..types.common import PagedResponse

class MarketplaceResource:
    """Marketplace domain — Sync.
    
    @param base - marketplaces v2 base URL (`{gateway}/v2/backend/marketplaces`)
    """

    def __init__(self, http: HttpTransport, base: str):
        self._http = http
        self._marketplaces = base.rstrip("/")
        # templates are usually at the same level as marketplaces in v2/backend
        self._templates = self._marketplaces.replace("/marketplaces", "/templates")

    # --- Marketplace templates ---
    def list_use_case_templates(self) -> List[Dict[str, Any]]:
        return self._http.request("GET", f"{self._marketplaces}/templates").json()

    def install_from_json(self, body: Dict[str, Any]) -> Dict[str, Any]:
        """Install a template from JSON definition. Endpoint: /templates/v2/custom"""
        return self._http.request("POST", f"{self._marketplaces}/templates/v2/custom", json=body).json()

    # --- Marketplace products ---
    def list_products(self, page: Optional[int] = None, limit: Optional[int] = None, 
                      category: Optional[str] = None, search: Optional[str] = None) -> PagedResponse[Product]:
        params = {}
        if page: params["page"] = page
        if limit: params["limit"] = limit
        if category: params["category"] = category
        if search: params["search"] = search
        return self._http.request("GET", f"{self._marketplaces}/products", params=params).json()

    def get_product(self, product_id: str) -> Product:
        return self._http.request("GET", f"{self._marketplaces}/products/{product_id}").json()

    def install_product(self, product_id: str) -> Order:
        """Quick install a product by ID."""
        return self._http.request("POST", f"{self._marketplaces}/installations/{product_id}").json()

    # --- Orders ---
    def create_order(self, body: CreateOrderInput) -> Order:
        product_id = body.get("product_id", "")
        return self._http.request("POST", f"{self._marketplaces}/installations/{product_id}", json=body).json()

    def list_orders(self, page: Optional[int] = None, limit: Optional[int] = None, 
                    status: Optional[str] = None) -> PagedResponse[Order]:
        params = {}
        if page: params["page"] = page
        if limit: params["limit"] = limit
        if status: params["status"] = status
        return self._http.request("GET", f"{self._marketplaces}/orders", params=params).json()

    def get_order(self, order_id: str) -> Order:
        return self._http.request("GET", f"{self._marketplaces}/orders/{order_id}").json()

    def update_order_status(self, order_id: str, status: str) -> Order:
        return self._http.request("PATCH", f"{self._marketplaces}/orders/{order_id}/status", json={"status": status}).json()

    # --- Files ---
    def upload_file(self, files: Any) -> MarketplaceFileUploadResponse:
        return self._http.request("POST", f"{self._marketplaces}/files", files=files).json()

    def delete_file(self, file_id: str) -> None:
        self._http.request("DELETE", f"{self._marketplaces}/files/{file_id}")

    def get_file_details(self, file_id: str) -> MarketplaceFileDetails:
        return self._http.request("GET", f"{self._marketplaces}/file-details/{file_id}").json()

    def download_file(self, short_path: str) -> Any:
        return self._http.request("GET", f"{self._marketplaces}/download/{short_path}")

    # --- Email Templates ---
    def list_email_templates(self, params: Optional[Dict[str, str]] = None) -> List[EmailTemplate]:
        return self._http.request("GET", f"{self._marketplaces}/email-templates/search", params=params or {}).json()

    def create_email_template(self, body: CreateEmailTemplateInput) -> EmailTemplate:
        return self._http.request("POST", f"{self._marketplaces}/email-templates", json=body).json()

    def post_channel_workflows(self, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._marketplaces}/channel-workflows", json=body).json()

    # --- Categories ---
    def list_categories(self, params: Optional[Dict[str, str]] = None) -> List[MarketplaceCategory]:
        return self._http.request("GET", f"{self._marketplaces}/categories", params=params or {}).json()


class AsyncMarketplaceResource:
    """Marketplace domain — Async."""

    def __init__(self, http: AsyncHttpTransport, base: str):
        self._http = http
        self._marketplaces = base.rstrip("/")
        self._templates = self._marketplaces.replace("/marketplaces", "/templates")

    async def list_use_case_templates(self) -> List[Dict[str, Any]]:
        res = await self._http.request("GET", f"{self._marketplaces}/templates")
        return res.json()

    async def install_from_json(self, body: Dict[str, Any]) -> Dict[str, Any]:
        res = await self._http.request("POST", f"{self._marketplaces}/templates/v2/custom", json=body)
        return res.json()

    async def list_products(self, page: Optional[int] = None, limit: Optional[int] = None, 
                          category: Optional[str] = None, search: Optional[str] = None) -> PagedResponse[Product]:
        params = {}
        if page: params["page"] = page
        if limit: params["limit"] = limit
        if category: params["category"] = category
        if search: params["search"] = search
        res = await self._http.request("GET", f"{self._marketplaces}/products", params=params)
        return res.json()

    async def get_product(self, product_id: str) -> Product:
        res = await self._http.request("GET", f"{self._marketplaces}/products/{product_id}")
        return res.json()

    async def install_product(self, product_id: str) -> Order:
        res = await self._http.request("POST", f"{self._marketplaces}/installations/{product_id}")
        return res.json()

    async def create_order(self, body: CreateOrderInput) -> Order:
        product_id = body.get("product_id", "")
        res = await self._http.request("POST", f"{self._marketplaces}/installations/{product_id}", json=body)
        return res.json()

    async def list_orders(self, page: Optional[int] = None, limit: Optional[int] = None, 
                        status: Optional[str] = None) -> PagedResponse[Order]:
        params = {}
        if page: params["page"] = page
        if limit: params["limit"] = limit
        if status: params["status"] = status
        res = await self._http.request("GET", f"{self._marketplaces}/orders", params=params)
        return res.json()

    async def get_order(self, order_id: str) -> Order:
        res = await self._http.request("GET", f"{self._marketplaces}/orders/{order_id}")
        return res.json()

    async def update_order_status(self, order_id: str, status: str) -> Order:
        res = await self._http.request("PATCH", f"{self._marketplaces}/orders/{order_id}/status", json={"status": status})
        return res.json()

    async def list_categories(self, params: Optional[Dict[str, str]] = None) -> List[MarketplaceCategory]:
        res = await self._http.request("GET", f"{self._marketplaces}/categories", params=params or {})
        return res.json()

File: ts/src/resources/marketplace.ts
import { HttpTransport } from "../http.js"
import type { Product, Order, CreateOrderInput, OrderStatus, PagedResponse } from "../types/index.js"

// ─── Email template interfaces ────────────────────────────────────────────────

export interface EmailTemplate {
  _id: string
  name?: string
  subject?: string
  body?: string
  [key: string]: unknown
}

export interface CreateEmailTemplateInput {
  name: string
  subject?: string
  body?: string
  [key: string]: unknown
}

// ─── Channel workflow interfaces ──────────────────────────────────────────────

export interface ChannelWorkflowInput {
  channel_id?: string
  workflow_id?: string
  [key: string]: unknown
}

export interface ChannelWorkflowResponse {
  success: boolean
  [key: string]: unknown
}

// ─── Install from JSON interface ──────────────────────────────────────────────

export interface InstallFromJsonInput {
  template?: Record<string, unknown>
  [key: string]: unknown
}

export interface InstallFromJsonResponse {
  success: boolean
  [key: string]: unknown
}

// ─── File interfaces ──────────────────────────────────────────────────────────

export interface MarketplaceFileDetails {
  _id: string
  name?: string
  url?: string
  [key: string]: unknown
}

export interface MarketplaceFileUploadResponse {
  url: string
  file_id?: string
  [key: string]: unknown
}

// ─── Category interface ───────────────────────────────────────────────────────

export interface MarketplaceCategory {
  _id: string
  name: string
  [key: string]: unknown
}

export class MarketplaceResource {
  /**
   * @param marketplacesBase - Standalone marketplaces service (gateway/marketplaces/v1)
   * @param platformBase     - Platform service base (gateway/platform) for platform/v2/marketplaces
   */
  constructor(
    private readonly http: HttpTransport,
    private readonly marketplacesBase: string,
    private readonly platformBase: string,
  ) {}

  private get standaloneV1() { return `${this.marketplacesBase}/v1` }
  private get platformV2() { return `${this.platformBase}/v2` }

  // ─── Standalone marketplaces service (/marketplaces/v1) ─────────────────────

  async listUseCaseTemplates(): Promise<{ _id: string; name?: string; [key: string]: unknown }[]> {
    return this.http.getFetch()(`${this.standaloneV1}/market-places/templates`, { method: "GET" }).then(r => r.json())
  }

  async installFromJson(body: InstallFromJsonInput): Promise<InstallFromJsonResponse> {
    return this.http.getFetch()(`${this.standaloneV1}/market-places/templates/install-from-json`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  // ─── Platform marketplaces (/platform/v2/marketplaces) ──────────────────────

  async listProducts(params?: {
    page?: number
    limit?: number
    category?: string
    search?: string
  }): Promise<PagedResponse<Product>> {
    const url = new URL(`${this.platformV2}/marketplaces/products`)
    if (params?.page)     url.searchParams.set("page",     String(params.page))
    if (params?.limit)    url.searchParams.set("limit",    String(params.limit))
    if (params?.category) url.searchParams.set("category", params.category)
    if (params?.search)   url.searchParams.set("search",   params.search)
    return this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
  }

  async getProduct(productId: string): Promise<Product> {
    return this.http.getFetch()(`${this.platformV2}/marketplaces/products/${productId}`, { method: "GET" }).then(r => r.json())
  }

  async installProduct(productId: string): Promise<Order> {
    return this.http.getFetch()(`${this.platformV2}/marketplaces/installations/${productId}`, {
      method: "POST",
    }).then(r => r.json())
  }

  async createOrder(body: CreateOrderInput): Promise<Order> {
    return this.http.getFetch()(`${this.platformV2}/marketplaces/installations/${body.product_id ?? ""}`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async listOrders(params?: { page?: number; limit?: number; status?: OrderStatus }): Promise<PagedResponse<Order>> {
    const url = new URL(`${this.platformV2}/marketplaces/orders`)
    if (params?.page)   url.searchParams.set("page",   String(params.page))
    if (params?.limit)  url.searchParams.set("limit",  String(params.limit))
    if (params?.status) url.searchParams.set("status", params.status)
    return this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
  }

  async getOrder(orderId: string): Promise<Order> {
    return this.http.getFetch()(`${this.platformV2}/marketplaces/orders/${orderId}`, { method: "GET" }).then(r => r.json())
  }

  async updateOrderStatus(orderId: string, status: OrderStatus): Promise<Order> {
    return this.http.getFetch()(`${this.platformV2}/marketplaces/orders/${orderId}/status`, {
      method: "PATCH",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ status }),
    }).then(r => r.json())
  }

  // ─── Files ───────────────────────────────────────────────────────────────────

  async uploadFile(body: FormData): Promise<MarketplaceFileUploadResponse> {
    return this.http.getFetch()(`${this.platformV2}/marketplaces/files`, {
      method: "POST",
      body,
    }).then(r => r.json())
  }

  async deleteFile(fileId: string): Promise<void> {
    await this.http.getFetch()(`${this.platformV2}/marketplaces/files/${fileId}`, { method: "DELETE" })
  }

  async getFileDetails(fileId: string): Promise<MarketplaceFileDetails> {
    return this.http.getFetch()(`${this.platformV2}/marketplaces/file-details/${fileId}`, { method: "GET" }).then(r => r.json())
  }

  // ─── Email Templates ─────────────────────────────────────────────────────────

  async listEmailTemplates(params?: Record<string, string>): Promise<EmailTemplate[]> {
    const url = new URL(`${this.platformV2}/marketplaces/email-templates/search`)
    if (params) Object.entries(params).forEach(([k, v]) => url.searchParams.set(k, v))
    return this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
  }

  async createEmailTemplate(body: CreateEmailTemplateInput): Promise<EmailTemplate> {
    return this.http.getFetch()(`${this.platformV2}/marketplaces/email-templates`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async postChannelWorkflows(body: ChannelWorkflowInput): Promise<ChannelWorkflowResponse> {
    return this.http.getFetch()(`${this.platformV2}/marketplaces/channel-workflows`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async downloadMarketPlaceFile(shortPath: string): Promise<Response> {
    return this.http.getFetch()(`${this.platformV2}/marketplaces/download/${shortPath}`, { method: "GET" })
  }

  // ─── Categories ──────────────────────────────────────────────────────────────

  async listCategories(params?: Record<string, string>): Promise<MarketplaceCategory[]> {
    const url = new URL(`${this.platformV2}/marketplaces/categories`)
    if (params) Object.entries(params).forEach(([k, v]) => url.searchParams.set(k, v))
    return this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
  }
}
